### DISCUSSION TITLE
Propagate headers for traceability with OpenTelemetry

### DISCUSSION THREAD

**Discussion by nataliadamilano** [fastapi/fastapi#10514] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from Traceability.TracingService.TracingService import TracingService
from starlette.types import Receive, Send, Message, Scope
from starlette.types import ASGIApp
import six

class OpentelemetryMiddleware():

    def __init__(self, app: ASGIApp, tracing_service: TracingService) -> None:
        self.app = app
        self.tracing_service = tracing_service
        self.propagator_first = tracing_service.get_textmap_propagator()
        self.propagator_second = tracing_service.get_jaeger_propagator()

    async def __call__(self, scope: Scope, receive: Receive, send: Send):
        with self.tracing_service.get_span_manager() as span:

            async def send_wrapper(message: Message):
                if message["type"] == "http.response.start":
                    addHeaders = {}
                    self.propagator_first.inject(addHeaders)
                    self.propagator_second.inject(addHeaders)
                    for key,value in six.iteritems(addHeaders):
                        message['headers'].append( (key.encode(), value.encode()) )

                return await send(message)
            await self.app(scope, receive, send_wrapper)
```

### Description

- Adding middleware that propagates the tracing headers but does not do the job correctly. 
The propagation of the tracing headers is done through the request headers instead of the response headers as I am doing now. 
I understand that the request headers are immutable in FastAPI? Is this the case?
How can I solve this observability implementation?

### Operating System

Windows

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

3.8.1

**Comment by jgould22** [comment]
Why not look at an existing implementation ?

https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-fastapi/src/opentelemetry/instrumentation/fastapi/__init__.py

**Reply by nataliadamilano** [reply]
I want to add Traceparent and Uber-Trace-ID as request headers for propagation and I don't see that implementation on that resource.

**Reply by jgould22** [reply]
When you say you want to propagate them do you mean down stream from fastapi or back to the requesting client ?

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean and organized.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
